12-offer-encoding.md

```diff
@@ -238,296 +233,332 @@ The human-readable prefix for offers is `lno`.
-      - MUST set that minimum in `quantity_min`
+      - MUST set that minimum in `offer_quantity_min`
     - if the maximum quantity is known:
-      - MUST set that maximum in `quantity_max`
+      - MUST set that maximum in `offer_quantity_max`
     - if neither:
-      - MUST set `quantity_min` to 1 to indicate `quantity` is supported.
+      - MUST set `offer_quantity_min` to 1 to indicate `quantity` is supported.
     - if both:
-      - MUST set `quantity_min` less than or equal to `quantity_max`.
-    - MUST NOT set `quantity_min` or `quantity_max` less than 1.
-  - if `send_invoice` is present:
-    - if the offer is for a partial or full refund for a previously-paid
-      invoice:
-      - SHOULD set `refunded_payment_hash` to the `payment_hash` of that
-        invoice.
-  - otherwise:
-    - MUST NOT set `refunded_payment_hash`.
+      - MUST set `offer_quantity_min` less than or equal to `offer_quantity_max`.
+    - MUST NOT set `offer_quantity_min` or `offer_quantity_max` less than 1.
 
 A reader of an offer:
-  - if `features` contains unknown _odd_ bits that are non-zero:
+  - if the offer contains any unknown TLV fields greater or equal to 80:
+    - MUST NOT respond to the offer.
+  - if `offer_features` contains unknown _odd_ bits that are non-zero:
     - MUST ignore the bit.
-  - if `features` contains unknown _even_ bits that are non-zero:
+  - if `offer_features` contains unknown _even_ bits that are non-zero:
     - MUST NOT respond to the offer.
     - SHOULD indicate the unknown bit to the user.
-  - if `description` is not set:
+  - if `offer_description` is not set:
     - MUST NOT respond to the offer.
-  - if `node_id` is not set:
-    - if `paths` is not set or does not contain at least one `blinded_path`:
+  - if `offer_node_id` is not set:
+    - if `offer_paths` is not set or does not contain at least one `blinded_path`:
       - MUST NOT respond to the offer.
     - MUST use the final `onionmsg_path` `point` in the first `blinded_path` as the `node_id`.
-  - if `signature` is present, but is not a valid signature using `node_id` as described in [Signature Calculation](#signature-calculation):
-    - MUST NOT respond to the offer.
-  - if it uses `amount` to provide the user with a cost estimate:
+  - if it uses `offer_amount` to provide the user with a cost estimate:
     - MUST warn user if amount of actual invoice differs significantly
         from that expectation.
   - SHOULD not respond to an offer if the current time is after
-    `absolute_expiry`.
+    `offer_absolute_expiry`.
   - FIXME: more!
 
 ## Rationale
 
-A signature is optional because it makes for a longer string (potentially
+The entire offer is reflected in the invoice_request, both for
+completeness (so all information will be returned in the invoice), and
+so that the offer node can be stateless.  This makes `offer_metadata`
+particularly useful, since it can contain an authentication cookie to
+validate the other fields.
```

Comment:
I wanted to shave a byte off the QR code, sure.  But invoice / invoice_request space is way less expensive?  And offer_metadata is really only for stateless invoices; I don't expect to use it myself.

Full reflection is actually a nice simplification, since we did per-field reflection before.  (Disclaimer: I haven't *implemented* it yet!).